Enable ESLint JSDoc checks

colinrotherham opened the pull request:
I've enabled ESLint JSDoc checks in this PR

Links in with https://github.com/alphagov/govuk-frontend/issues/2884

We do have quite a few issues flagged but also:

1. ~JSDoc missing entirely (is this alright?)~
2. ~JSDoc wrongly flagging `@jest-environment` comments~

Have a look at the automated checks either way

**Update 1:** I’ve enabled rules that cover (I think) the “style” we’d like to use and fixed errors in places. ~Warnings are still logged but we can fix those in our own time.~

**Update 2:** All warnings now fixed

colinrotherham commented:
@romaricpascal I've improved the config to:

1. Make JSDoc blocks optional or "description only"
2. Require a hyphen between param and description ([matches TSDoc style](https://tsdoc.org/pages/tags/param/))
3. Allow the unknown return type `NodeListOf` for `.querySelectorAll()`
4. Allow the unknown tag `@jest-environment`

All the warnings should be genuine ones now

colinrotherham commented on .eslintrc.js:
```diff
@@ -11,6 +14,46 @@ module.exports = {
+      },
+      settings: {
+        jsdoc: {
+          mode: 'typescript'
```
Why `'typescript'` mode?

Rather than spend time documenting every complex object (or instance) we can use the declarations already published by the package author:

```mjs
/**
 * @param {import('puppeteer').Page} page
 * @param {import('puppeteer').ElementHandle} element
 * @param {import('puppeteer').WaitForOptions} options
 */
```

TypeScript declarations are widely available unlike local-only `@typedef` tags:

<img width="509" alt="Puppeteer on npmjs org showing TypeScript declarations are available" src="https://user-images.githubusercontent.com/415517/195338936-6aae0485-3c49-4bf7-9daf-dacd3fcb12c9.png">

colinrotherham commented on .eslintrc.js:
```diff
@@ -11,6 +14,46 @@ module.exports = {
+
+        // Require hyphens before param description
+        // Aligns with TSDoc style: https://tsdoc.org/pages/tags/param/
+        'jsdoc/require-hyphen-before-param-description': 'warn',
```
We sometimes hyphenate, sometimes don't

It's [recommended by JSDoc](https://jsdoc.app/tags-param.html) (and mandatory for TSDoc)

>If you provide a description, you can make the JSDoc comment more readable by inserting a hyphen before the description. Be sure to include a space before and after the hyphen.

colinrotherham commented on .eslintrc.js:
```diff
@@ -11,6 +14,46 @@ module.exports = {
   ],
   overrides: [
+    {
+      extends: 'plugin:jsdoc/recommended',
```
Well spotted! I thought I'd moved it, as it felt better in an override versus the root

For example if we added `*.ts` files they'd have their own override with TSDoc [via `eslint-plugin-tsdoc` instead](https://github.com/microsoft/tsdoc/tree/main/eslint-plugin#eslint-plugin-tsdoc)

So here I've made sure I'm only loading the JSDoc plugin for `'**/*.{cjs,js,mjs}'`

colinrotherham commented on .eslintrc.js:
```diff
@@ -11,6 +14,46 @@ module.exports = {
+        // Add missing .querySelectorAll() type
+        'jsdoc/no-undefined-types': [
+          'error', {
+            definedTypes: ['NodeListOf']
```
@romaricpascal Yeah exactly

Without this change then ~`NodeList<Element>`~ `NodeListOf<Element>` would be invalid

One of the JSDoc plugin dependencies must not be aware of it as a known DOM type (possibly because the TypeScript flavour is `NodeListOf` not `NodeList`. Haven't managed to track down exactly where to log this yet.

In Visual Studio Code if you mouse over `.querySelectorAll()` you'll see `NodeListOf<Element>`

<img width="789" alt="Screenshot of querySelectorAll being typed as NodeListOf" src="https://user-images.githubusercontent.com/415517/195412440-0432b551-8f04-4d96-b1b5-f78b638d0280.png">

colinrotherham commented on src/govuk/i18n.mjs:
```diff
@@ -204,15 +204,15 @@ I18n.prototype.getPluralSuffix = function (count) {
  * French: Armenian (hy), Bangla (bn), French (fr), Gujarati (gu), Hindi (hi),
- *   Persian Farsi (fa), Punjabi (pa), Zulu (zu)
+ * Persian Farsi (fa), Punjabi (pa), Zulu (zu)
  * German: Afrikaans (af), Albanian (sq), Azerbaijani (az), Basque (eu),
- *   Bulgarian (bg), Catalan (ca), Danish (da), Dutch (nl), English (en),
- *   Estonian (et), Finnish (fi), Georgian (ka), German (de), Greek (el),
- *   Hungarian (hu), Luxembourgish (lb), Norwegian (no), Somali (so),
- *   Swahili (sw), Swedish (sv), Tamil (ta), Telugu (te), Turkish (tr),
- *   Urdu (ur)
+ * Bulgarian (bg), Catalan (ca), Danish (da), Dutch (nl), English (en),
+ * Estonian (et), Finnish (fi), Georgian (ka), German (de), Greek (el),
+ * Hungarian (hu), Luxembourgish (lb), Norwegian (no), Somali (so),
+ * Swahili (sw), Swedish (sv), Tamil (ta), Telugu (te), Turkish (tr),
+ * Urdu (ur)
```
@romaricpascal I'll have a look. Wonder if we need to specify markdown with extra formatting?

colinrotherham replied:
I've restored the previous lines and everything seems fine now

I must have run `eslint --fix` with some stricter settings earlier 🤦‍♂️

colinrotherham commented on .eslintrc.js:
```diff
@@ -11,6 +14,46 @@ module.exports = {
   overrides: [
+    {
+      extends: 'plugin:jsdoc/recommended',
+      files: ['**/*.{cjs,js,mjs}'],
```
Yeah I'm happy with JSDoc checking in all JavaScript files 😮

colinrotherham commented on src/govuk/common.mjs:
```diff
@@ -6,6 +6,10 @@ import './vendor/polyfills/Element/prototype/closest.mjs'
  * This seems to fail in IE8, requires more investigation.
  * See: https://github.com/imagitama/nodelist-foreach-polyfill
+ *
+ * @param {NodeListOf<Element>} nodes - NodeList from querySelectorAll()
```
That's just me checking where it's called and could only see `NodeListOf<Element>`

Suppose you could have `document.querySelector('#example').childNodes` as `NodeListOf<ChildNode>`, shall I put both in the JSDoc? It's to avoid linting errors in TypeScript projects complaining that types are incompatible

```suggestion
 * @param {NodeListOf<Element | ChildNode>} nodes - NodeList from querySelectorAll()
```

colinrotherham commented on src/govuk/common.mjs:
```diff
@@ -6,6 +6,10 @@ import './vendor/polyfills/Element/prototype/closest.mjs'
  * See: https://github.com/imagitama/nodelist-foreach-polyfill
+ *
+ * @param {NodeListOf<Element>} nodes - NodeList from querySelectorAll()
+ * @param {(value: Element, key: number, parent: NodeListOf<Element>) => void} callback - Callback function to run for each node
```
Where's this JSDoc block from @romaricpascal?

colinrotherham replied:
Ahh I see, that's your suggestion

Because we've set `mode: 'typescript'` for compatibility with **node_modules** that export type declarations, maybe for the `jsdoc` CLI you'll also need to enable the TypeScript plugin?

But it's simple enough to use `@callback` so pushed another commit up 👍

colinrotherham replied:
Forgot to update this comment

I've added **JSDoc generation** via https://github.com/alphagov/govuk-frontend/pull/2920 to confirm it all works

Type declarations like `import('Puppeteer').Page` are handled by the [`jsdoc-tsimport-plugin`](https://www.npmjs.com/package/jsdoc-tsimport-plugin) plugin

colinrotherham replied:
Ahh sorry @36degrees that was @romaricpascal's suggestion too, it's a `@callback` now 🙌 

To answer your question though I grabbed the names from Microsoft's `forEach` definition:

https://github.com/microsoft/TypeScript/blob/main/lib/lib.dom.d.ts#L10079
```ts
interface NodeListOf<TNode extends Node> extends NodeList {
    item(index: number): TNode;
    /**
     * Performs the specified action for each node in an list.
     * @param callbackfn  A function that accepts up to three arguments. forEach calls the callbackfn function one time for each element in the list.
     * @param thisArg  An object to which the this keyword can refer in the callbackfn function. If thisArg is omitted, undefined is used as the this value.
     */
    forEach(callbackfn: (value: TNode, key: number, parent: NodeListOf<TNode>) => void, thisArg?: any): void;
    [index: number]: TNode;
}
```

colinrotherham commented:
> One thing I'm still unclear is where the line is regarding the safety we're looking to get from the JSDoc linting ahead of the release. Is it that the JSDoc we produce can be consumed as type definitions by VSCode?

Thanks @romaricpascal

In future maybe, but not to generate type definitions yet. We need to remember that types will be automatically _inferred_ from the JSDoc so we'll want to make sure they're all correct—eventually

I added some information on this issue: https://github.com/alphagov/govuk-frontend/issues/2835#issuecomment-1276431852

romaricpascal commented:
Having the rules set to warn rather than error would allow us to merge that configuration quickly, however I'm worried we won't have much of a gain from it. The test logs will just get filled with the current breakages and I'm not sure who will go there to fix them.

Having them as errors would force us to either not output JSDoc, or when we do it, do it to a consistent standard (even if we set the bar to a lower level than the plugin's recommended settings). It'd make more sense to me that this PR introducing linting does it to raise errors rather than risk the errors just laying around in the background, increasing the complexity of our linting for not much of a gain in the quality of our docs (besides our editor autofixing things on save).

I would welcome raising errors and I'm happy to go around and fix the issues so we can start on a clean codebase. I'm not strongly attached to it, though and if people would prefer a softer introduction of this linting, then we could keep it as warnings. I'll let other @alphagov/design-system-developers voice their thoughts 😄

colinrotherham commented:
@romaricpascal I've fixed and added a minimum amount of JSDoc to all the remaining files

No more warnings logged so we can bring back `--max-warnings 0`

colinrotherham commented:
> One (non blocking) little **question**: What's the difference between configuring rules as `error` and using `--max-warnings 0`?

Mainly for code editor and IDE integration

* Warnings appear in yellow (less scary)
* Errors appear in red (more scary)

But adding `--max-warnings 0` makes warnings break the build like errors do

36degrees commented on src/govuk/common.mjs:
```diff
@@ -34,18 +40,24 @@ export function generateUniqueID () {
  *
- * @param {...Object} - Any number of objects to merge together.
- * @returns {Object} - A flattened object of key-value pairs.
+ * @returns {object} A flattened object of key-value pairs.
```
Just want to check my understanding of the removal of `@param` – presumably ESLint doesn't like the fact that we had `@param` annotations for params that weren't in the actual function signature itself, and we can't fix that at the minute because we can't use the spread operator here?

colinrotherham replied:
@36degrees Yeah, believe me I tried 😭 

Yeah, it doesn't like it because there literally is no `@param`. We use [`arguments`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/arguments) instead.

Did you see I'd added a comment in place of the function params?
(Thought best to try to explain it another way at least)

```mjs
export function mergeConfigs (/* configObject1, configObject2, ...configObjects */) {
```

Parameter spread `(...args)` can be documented in future though

For backwards compatibility we could add Babel to Rollup. It'd pick up the `@babel/plugin-transform-parameters` plugin based on our [**.browserslistrc**](https://github.com/alphagov/govuk-frontend/blob/main/.browserslistrc) config. Transforming `(...args)` → `arguments`

They show before/after examples here:
https://babeljs.io/docs/en/babel-plugin-transform-parameters

36degrees replied:
Let's live with this for now and we can revisit when we look at introducing Babel more widely 👍🏻

36degrees replied:
Unless we want to use an `/* eslint-disable no-alert, no-console */` so we can add the 'correct' JSDoc without the linter complaining about it?

colinrotherham replied:
We could, but then the `jsdoc` CLI would complain in this PR https://github.com/alphagov/govuk-frontend/pull/2920

Just end up kicking the problem from tool to tool unfortunately

36degrees commented on src/govuk/common.mjs:
```diff
@@ -192,8 +204,8 @@ export function normaliseDataset (dataset) {
- * @param {String} attributeName - The name of the attribute
- * @returns {String|undefined}
+ * @param {string} attributeName - The name of the attribute
+ * @returns {string | null} Attribute value
```
Although getAttribute can return `null`, I don't think it's a logical possibility here because we've used an attribute selector which means the attribute _must_ exist.

However, if there is no element with that attribute then there is no explicit return from the function which I think means it would return `undefined` rather than `null`?

colinrotherham replied:
I do appreciate the through reviews, thank you for this. I've got into a habit of using the TypeScript language service to check the return types everywhere which flagged this up

Like you say, it's correct that **getAttribute()** returns `string | null`, but it's not clever enough to know we're selecting the same attribute above 🤦‍♂️ 

I think JSDoc 3.2+ supports `{string | void}` (nothing returned) but I'll put `undefined` back as it's still valid

```suggestion
 * @returns {string | undefined} Attribute value
```

36degrees commented on src/govuk/i18n.mjs:
```diff
@@ -263,8 +263,8 @@ I18n.pluralRulesMap = {
  *
- * @param    {number}  n  - The `count` number being passed through. This must be a positive integer. Negative numbers and decimals aren't accounted for.
- * @returns  {string}     - The string that needs to be suffixed to the key (without separator).
+ * @param {number} n - The `count` number being passed through. This must be a positive integer. Negative numbers and decimals aren't accounted for.
+ * @returns {string} The string that needs to be suffixed to the key (without separator).
```
Not 100% sure that this is the right way to document `I18n.pluralRules` (it's an object containing multiple functions, rather than a function that takes params itself)

(ES Lint doesn't seem to mind… so could look at fixing this later)

36degrees commented on .eslintrc.js:
```diff
@@ -11,6 +11,46 @@ module.exports = {
+      },
+      settings: {
+        jsdoc: {
+          mode: 'typescript'
```
Ah, I was reviewing commit by commit so didn't see that. I think it might be worth capturing as a comment in the config file? Perhaps:

```suggestion
         // Allows us to re-use declarations that exist in our dependencies
          mode: 'typescript'
```

colinrotherham commented on .eslintrc.js:
```diff
@@ -11,6 +11,46 @@ module.exports = {
+        // JSDoc blocks are optional
+        'jsdoc/require-jsdoc': 'off',
+        'jsdoc/require-param-description': 'off',
+        'jsdoc/require-param': 'off',
```
@36degrees Yeah if we solely have `'jsdoc/require-jsdoc'` it'll do that

ESLint is clever enough to make `@returns` optional if nothing is returned, but we also needed `@description` and `@param` optional for things like this:

```mjs
/**
 * Umbrella scripts tasks (for watch)
 * Runs JavaScript code quality checks and compilation
 */
gulp.task('scripts', gulp.series(
  npmScriptTask('lint:js', ['--silent']),
  'js:compile'
))
```

☝️ We quite frequently use JSDoc style comments for scripts and tests

We have a few other options on the table:

1. _Keep_ these rules as-is with JSDoc block `@param` and `@description` optional
2. _Remove_ these rules and lint for JSDoc blocks without `@param` and `@description`
3. _Remove_ these rules BUT exclude **./src**
